test/contrasts.jl

```diff
@@ -81,6 +81,17 @@
                                 1 -1 -1
                                 1  0  1]
     @test coefnames(mf) == ["(Intercept)"; "x: c"; "x: b"]
+    
+    # Ordinal factors
+    
+    data = DataFrame(x = levels!(categorical(['A', 'B', 'C', 'C'],
+                                             ordered = true),
+                                 ['C', 'B', 'A']))
+    f = @formula(x ~ 1)
+    f = apply_schema(f, schema(data))
+    y = modelcols(f.lhs, data)
+    f = apply_schema(f, schema(data, Dict(:x => DummyCoding(levels = levels(data.x)))))
```

Comment:
Passing `levels` here is redundant and defeats the point of using a categorical array, right?

Reply:
It is redundant now, but works as a test since it errors with current master. That is just to verify that it is working as if/intended.

Reply:
OK but then below you need to check that `y == modelcols(f.lhs, data) == somereference` (or call `@test y == somereference` twice).

Reply:
Ah, and also better pass levels in a different order from `levels` and from the order of appearance, to ensure it really has an effect.

Reply:
And add an unused value to `levels` to check that it's dropped.

Reply:
That is what it is doing, it is first computing the value and storing it as `y` and the I test the value against the reference which is `modelcols(f.lhs, data)` with the formula now been updated with an explicit order by contrasts.

Reply:
The test for unused levels was already included in the `modelmatrix.jl`. I put a comment on that test that it also checks correctness against non-occurring levels.

Reply:
> That is what it is doing, it is first computing the value and storing it as `y` and the I test the value against the reference which is `modelcols(f.lhs, data)` with the formula now been updated with an explicit order by contrasts.

Yes but this doesn't test the fact that passing levels to `DummyCoding` has any effect. Better check against an explicit expected array.